prereise/gather/winddata/hrrr/calculations.py

```diff
@@ -0,0 +1,89 @@
+
+from scipy.spatial import KDTree
+
+from prereise.gather.winddata.rap.power_curves import (
+    get_power,
+    get_state_power_curves,
+    get_turbine_power_curves,
+)
+
+U_COMPONENT_SELECTOR="U component of wind"
+V_COMPONENT_SELECTOR="V component of wind"
+
+
+def get_wind_data_lat_long(start_dt, directory):
+    gribs = pygrib.open(directory + formatted_filename(start_dt, "sfc", 0))
+    grib = next(gribs)
+    return grib.latlons()
+
+
+def find_closest_wind_grids(wind_farms, wind_data_lat_long):
+    grid_lats, grid_lons = wind_data_lat_long[0].flatten(), wind_data_lat_long[1].flatten()
+    grid_lat_lon_length = len(grid_lats)
+    assert grid_lat_lon_length == len(grid_lons)
+    grid_lat_lon_unit_vectors = [ll2uv(grid_lons[i], grid_lats[i]) for i in range(grid_lat_lon_length)]
+
+    tree = KDTree(grid_lat_lon_unit_vectors)
+
+    wind_farm_lats = wind_farms.lat.values
+    wind_farm_lons = wind_farms.lon.values
+
+    wind_farm_ct = len(wind_farms)
+
+    wind_farm_unit_vectors = [ll2uv(wind_farm_lons[i], wind_farm_lats[i]) for i in range(wind_farm_ct)]
+    _, indices = tree.query(wind_farm_unit_vectors)
+
+    return indices
+
+
+def calculate_pout(wind_farms, start_dt, end_dt, directory):
+    # Returns pandas dataframe in the following format:
+    #       wind_farm1  wind_farm2
+    # dt1        1        2
+    # dt2        3        4
+
+    id_target = wind_farms.index.values
+    state_target = [
+        "Offshore"
+        if wind_farms.loc[i].type == "wind_offshore"
+        else id2abv[wind_farms.loc[i].zone_id]
+        for i in id_target
+    ]
+    n_target = len(wind_farms)
+
+    tpc = get_turbine_power_curves()
+    spc = get_state_power_curves()
+
+    wind_data_lat_long = get_wind_data_lat_long(start_dt, directory)
+
+    wind_farm_to_closest_wind_grid_indices = find_closest_wind_grids(wind_farms, wind_data_lat_long)
+
+    print(wind_farm_to_closest_wind_grid_indices)
+
+    dts = [dt for dt in date_range(start=start_dt, end=end_dt, freq="H").to_pydatetime()]
+    data = np.empty((len(dts), len(wind_farm_to_closest_wind_grid_indices)))
+
+    for i, dt in enumerate(dts):
+        gribs = pygrib.open(formatted_filename(dt, "sfc", hours_forecasted=0))
+        u_component = gribs.select(name=U_COMPONENT_SELECTOR)[0].values.flatten()
+        v_component = gribs.select(name=V_COMPONENT_SELECTOR)[0].values.flatten()
+        wind_farm_specific_u_component = u_component[wind_farm_to_closest_wind_grid_indices]
+        wind_farm_specific_v_component = v_component[wind_farm_to_closest_wind_grid_indices]
+        wind_farm_specific_magnitude = np.sqrt(pow(wind_farm_specific_u_component, 2) + pow(wind_farm_specific_v_component, 2))
+        
+        power = np.array([
+            get_power(tpc, spc, wind_farm_specific_magnitude[j], state_target[j])
+            for j in range(n_target)
+        ])
+        data[i] = power
+
+    df = pd.DataFrame(data=data, index=dts, columns=wind_farms.index.values)
```

Comment:
you can drop the `values` here too.

Reply:
Also, I guess you could do:
```
df = pd.DataFrame({i: 0 for i in wind_farms.index}, index=date_range(start=start_dt, end=end_dt, freq="H"))
for t in df.index:
    gribs = pygrib.open(formatted_filename(t.to_pydatetime()., "sfc", hours_forecasted=0))
    u_component = gribs.select(name=U_COMPONENT_SELECTOR)[0].values.flatten()
    v_component = gribs.select(name=V_COMPONENT_SELECTOR)[0].values.flatten()
    wind_farm_specific_u_component = u_component[wind_farm_to_closest_wind_grid_indices]
    wind_farm_specific_v_component = v_component[wind_farm_to_closest_wind_grid_indices]
    wind_farm_specific_magnitude = np.sqrt(pow(wind_farm_specific_u_component, 2) + pow(wind_farm_specific_v_component, 2))
    df.loc[t, :] = [get_power(tpc, spc, i, j) for i, j in zip(wind_farm_specific_magnitude, state_target)]
```

Reply:
I appreciate the suggestion but I think I'll keep it a bit more verbose as I think the current state is a bit more readable. The suggestion was a bit hard for me the follow and understand what the ultimate dataframe looked like where as 
`df = pd.DataFrame(data=data, index=dts, columns=wind_farms.index)` is very explicit and clear